File: knowledge/heuristic_loader.py
"""
HeuristicLoader - Load and manage UX heuristics from Baymard Institute and Nielsen Norman Group
"""

import json
import os
from typing import List, Dict, Any, Optional
from dataclasses import dataclass, asdict
from pathlib import Path
from loguru import logger
import numpy as np
from sentence_transformers import SentenceTransformer


@dataclass
class UXHeuristic:
    """Represents a single UX heuristic/guideline"""
    id: str
    source: str  # 'baymard' or 'nng'
    category: str
    title: str
    description: str
    context: List[str]  # e.g., ['checkout', 'forms', 'mobile']
    severity_weight: float  # 0.0-1.0
    persona_relevance: Dict[str, float]  # persona_name -> relevance score
    examples: List[str]
    citation_url: str


class HeuristicLoader:
    """
    Loads, manages, and retrieves UX heuristics with RAG-based semantic search.
    """
    
    def __init__(self, knowledge_dir: str = "knowledge"):
        self.knowledge_dir = Path(knowledge_dir)
        self.heuristics_file = self.knowledge_dir / "heuristics_db.json"
        self.embeddings_file = self.knowledge_dir / "heuristic_embeddings.npy"
        
        self.heuristics: List[UXHeuristic] = []
        self.embeddings: Optional[np.ndarray] = None
        self.embedding_model = SentenceTransformer('all-MiniLM-L6-v2')
        
        self._ensure_knowledge_dir()
        self._load_or_initialize()
    
    def _ensure_knowledge_dir(self):
        """Create knowledge directory if it doesn't exist"""
        self.knowledge_dir.mkdir(exist_ok=True)
    
    def _load_or_initialize(self):
        """Load existing heuristics or initialize with defaults"""
        if self.heuristics_file.exists():
            self._load_heuristics()
            logger.info(f"Loaded {len(self.heuristics)} heuristics from database")
        else:
            self._initialize_default_heuristics()
            self._save_heuristics()
            logger.info("Initialized default heuristics database")
    
    def _initialize_default_heuristics(self):
        """Initialize with curated Baymard + NN/g heuristics"""
        
        # Baymard Institute - E-commerce focused
        baymard_heuristics = [
            UXHeuristic(
                id="BAY-001",
                source="baymard",
                category="Touch Targets",
                title="Minimum Touch Target Size",
                description="Touch targets should be at least 44x44 CSS pixels to ensure easy tapping, especially for users with motor impairments or on mobile devices.",
                context=["mobile", "buttons", "links", "forms"],
                severity_weight=0.9,
                persona_relevance={
                    "senior_casual": 1.0,
                    "accessibility_focused": 1.0,
                    "middle_age_moderate": 0.7,
                    "young_power_user": 0.3
                },
                examples=[
                    "Button too small: 28x28px",
                    "Link spacing insufficient: 8px between targets"
                ],
                citation_url="https://baymard.com/blog/touch-target-size"
            ),
            UXHeuristic(
                id="BAY-002",
                source="baymard",
                category="Form Design",
                title="Inline Form Validation",
                description="Provide immediate, inline validation feedback as users complete form fields. Show success indicators for valid input and clear error messages for invalid input.",
                context=["forms", "checkout", "registration"],
                severity_weight=0.8,
                persona_relevance={
                    "senior_casual": 0.9,
                    "accessibility_focused": 0.8,
                    "middle_age_moderate": 0.8,
                    "young_power_user": 0.6
                },
                examples=[
                    "No validation until form submission",
                    "Error messages not near the problematic field"
                ],
                citation_url="https://baymard.com/blog/inline-form-validation"
            ),
            UXHeuristic(
                id="BAY-003",
                source="baymard",
                category="Checkout Flow",
                title="Guest Checkout Option",
                description="Always provide a guest checkout option. Forcing account creation increases cart abandonment by 23%.",
                context=["checkout", "e-commerce", "registration"],
                severity_weight=0.95,
                persona_relevance={
                    "senior_casual": 0.9,
                    "accessibility_focused": 0.6,
                    "middle_age_moderate": 0.8,
                    "young_power_user": 0.7
                },
                examples=[
                    "Forced account creation before checkout",
                    "No 'Continue as Guest' button visible"
                ],
                citation_url="https://baymard.com/blog/checkout-flow"
            ),
            UXHeuristic(
                id="BAY-004",
                source="baymard",
                category="Visual Hierarchy",
                title="Clear Visual Hierarchy",
                description="Use size, color, and spacing to create a clear visual hierarchy. Primary actions should be visually prominent.",
                context=["buttons", "navigation", "layout"],
                severity_weight=0.7,
                persona_relevance={
                    "senior_casual": 0.95,
                    "accessibility_focused": 0.9,
                    "middle_age_moderate": 0.8,
                    "young_power_user": 0.5
                },
                examples=[
                    "Primary and secondary buttons look identical",
                    "Important information buried in dense text"
                ],
                citation_url="https://baymard.com/blog/visual-hierarchy"
            ),
            UXHeuristic(
                id="BAY-005",
                source="baymard",
                category="Mobile Navigation",
                title="Thumb-Friendly Navigation",
                description="Place primary navigation and actions in the thumb zone (bottom 1/3 of screen) for one-handed mobile use.",
                context=["mobile", "navigation", "buttons"],
                severity_weight=0.8,
                persona_relevance={
                    "senior_casual": 0.8,
                    "accessibility_focused": 0.7,
                    "middle_age_moderate": 0.7,
                    "young_power_user": 0.9
                },
                examples=[
                    "Primary action at top of screen",
                    "Navigation requires two-handed use"
                ],
                citation_url="https://baymard.com/blog/mobile-navigation"
            )
        ]
        
        # Nielsen Norman Group - General UX
        nng_heuristics = [
            UXHeuristic(
                id="NNG-001",
                source="nng",
                category="Visibility of System Status",
                title="Keep Users Informed",
                description="The system should always keep users informed about what is going on through appropriate feedback within reasonable time.",
                context=["loading", "feedback", "progress"],
                severity_weight=0.85,
                persona_relevance={
                    "senior_casual": 0.9,
                    "accessibility_focused": 1.0,
                    "middle_age_moderate": 0.8,
                    "young_power_user": 0.6
                },
                examples=[
                    "No loading indicator during long operations",
                    "Form submission with no confirmation"
                ],
                citation_url="https://www.nngroup.com/articles/ten-usability-heuristics/"
            ),
            UXHeuristic(
                id="NNG-002",
                source="nng",
                category="Error Prevention",
                title="Prevent Errors Before They Occur",
                description="Good error messages are important, but better designs prevent problems from occurring in the first place.",
                context=["forms", "validation", "constraints"],
                severity_weight=0.9,
                persona_relevance={
                    "senior_casual": 1.0,
                    "accessibility_focused": 0.9,
                    "middle_age_moderate": 0.9,
                    "young_power_user": 0.7
                },
                examples=[
                    "No input constraints on date fields",
                    "Destructive actions without confirmation"
                ],
                citation_url="https://www.nngroup.com/articles/slips/"
            ),
            UXHeuristic(
                id="NNG-003",
                source="nng",
                category="Recognition vs Recall",
                title="Minimize Memory Load",
                description="Minimize the user's memory load by making objects, actions, and options visible. Instructions should be visible or easily retrievable.",
                context=["navigation", "forms", "instructions"],
                severity_weight=0.8,
                persona_relevance={
                    "senior_casual": 1.0,
                    "accessibility_focused": 0.9,
                    "middle_age_moderate": 0.9,
                    "young_power_user": 0.5
                },
                examples=[
                    "Hidden navigation requiring memorization",
                    "Multi-step process without progress indicator"
                ],
                citation_url="https://www.nngroup.com/articles/recognition-and-recall/"
            ),
            UXHeuristic(
                id="NNG-004",
                source="nng",
                category="Consistency",
                title="Follow Platform Conventions",
                description="Users should not have to wonder whether different words, situations, or actions mean the same thing. Follow platform conventions.",
                context=["navigation", "terminology", "patterns"],
                severity_weight=0.75,
                persona_relevance={
                    "senior_casual": 0.9,
                    "accessibility_focused": 0.8,
                    "middle_age_moderate": 0.8,
                    "young_power_user": 0.6
                },
                examples=[
                    "Inconsistent button placement across pages",
                    "Non-standard icons without labels"
                ],
                citation_url="https://www.nngroup.com/articles/consistency-and-standards/"
            ),
            UXHeuristic(
                id="NNG-005",
                source="nng",
                category="Accessibility",
                title="Sufficient Color Contrast",
                description="Text and interactive elements must have sufficient color contrast (WCAG AA: 4.5:1 for normal text, 3:1 for large text).",
                context=["accessibility", "visual", "text"],
                severity_weight=0.95,
                persona_relevance={
                    "senior_casual": 1.0,
                    "accessibility_focused": 1.0,
                    "middle_age_moderate": 0.9,
                    "young_power_user": 0.4
                },
                examples=[
                    "Light gray text on white background (2:1 contrast)",
                    "Low contrast buttons"
                ],
                citation_url="https://www.nngroup.com/articles/low-contrast/"
            ),
            UXHeuristic(
                id="NNG-006",
                source="nng",
                category="Flexibility",
                title="Accelerators for Expert Users",
                description="Provide accelerators (shortcuts, gestures) for experienced users while keeping the interface accessible to novices.",
                context=["keyboard", "shortcuts", "efficiency"],
                severity_weight=0.6,
                persona_relevance={
                    "senior_casual": 0.3,
                    "accessibility_focused": 0.8,
                    "middle_age_moderate": 0.5,
                    "young_power_user": 1.0
                },
                examples=[
                    "No keyboard shortcuts available",
                    "Cannot skip repetitive steps"
                ],
                citation_url="https://www.nngroup.com/articles/flexibility-efficiency-heuristic/"
            )
        ]
        
        self.heuristics = baymard_heuristics + nng_heuristics
        self._generate_embeddings()
    
    def _generate_embeddings(self):
        """Generate embeddings for all heuristics"""
        texts = [
            f"{h.title}. {h.description}. Context: {', '.join(h.context)}"
            for h in self.heuristics
        ]
        self.embeddings = self.embedding_model.encode(texts, convert_to_numpy=True)
        
        # Save embeddings
        np.save(self.embeddings_file, self.embeddings)
        logger.info(f"Generated embeddings for {len(self.heuristics)} heuristics")
    
    def get_relevant_guidelines(
        self,
        ui_context: str,
        persona_name: str = None,
        top_k: int = 5,
        min_similarity: float = 0.3
    ) -> List[Dict[str, Any]]:
        """
        Retrieve the most relevant UX guidelines for the current UI context.
        
        Args:
            ui_context: Description of current screen/page (e.g., "checkout payment form")
            persona_name: Name of persona to weight relevance
            top_k: Number of guidelines to return
            min_similarity: Minimum similarity threshold
            
        Returns:
            List of relevant guidelines with scores
        """
        # Generate embedding for query
        query_embedding = self.embedding_model.encode([ui_context], convert_to_numpy=True)[0]
        
        # Calculate cosine similarity
        similarities = np.dot(self.embeddings, query_embedding) / (
            np.linalg.norm(self.embeddings, axis=1) * np.linalg.norm(query_embedding)
        )
        
        # Get top-k indices
        top_indices = np.argsort(similarities)[::-1][:top_k * 2]  # Get more for filtering
        
        results = []
        for idx in top_indices:
            if similarities[idx] < min_similarity:
                continue
                
            heuristic = self.heuristics[idx]
            score = float(similarities[idx])
            
            # Apply persona weighting if specified
            if persona_name and persona_name in heuristic.persona_relevance:
                persona_weight = heuristic.persona_relevance[persona_name]
                weighted_score = score * (0.7 + 0.3 * persona_weight)  # 70% similarity, 30% persona
            else:
                weighted_score = score
            
            results.append({
                'heuristic': asdict(heuristic),
                'similarity_score': score,
                'weighted_score': weighted_score,
                'persona_relevance': heuristic.persona_relevance.get(persona_name, 0.5) if persona_name else None
            })
        
        # Sort by weighted score and return top-k
        results.sort(key=lambda x: x['weighted_score'], reverse=True)
        return results[:top_k]
    
    def get_by_category(self, category: str) -> List[UXHeuristic]:
        """Get all heuristics in a specific category"""
        return [h for h in self.heuristics if h.category.lower() == category.lower()]
    
    def get_by_context(self, context: str) -> List[UXHeuristic]:
        """Get all heuristics relevant to a specific context"""
        return [h for h in self.heuristics if context.lower() in [c.lower() for c in h.context]]
    
    def get_high_priority_for_persona(self, persona_name: str, min_relevance: float = 0.7) -> List[UXHeuristic]:
        """Get high-priority heuristics for a specific persona"""
        return [
            h for h in self.heuristics
            if h.persona_relevance.get(persona_name, 0) >= min_relevance
        ]
    
    def _save_heuristics(self):
        """Save heuristics to JSON file"""
        data = [asdict(h) for h in self.heuristics]
        with open(self.heuristics_file, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        logger.info(f"Saved {len(self.heuristics)} heuristics to {self.heuristics_file}")
    
    def _load_heuristics(self):
        """Load heuristics from JSON file"""
        with open(self.heuristics_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        self.heuristics = [UXHeuristic(**item) for item in data]
        
        # Load embeddings if available
        if self.embeddings_file.exists():
            self.embeddings = np.load(self.embeddings_file)
        else:
            self._generate_embeddings()
    
    def add_custom_heuristic(self, heuristic: UXHeuristic):
        """Add a custom heuristic to the database"""
        self.heuristics.append(heuristic)
        self._generate_embeddings()
        self._save_heuristics()
        logger.info(f"Added custom heuristic: {heuristic.id}")


# Example usage
if __name__ == "__main__":
    loader = HeuristicLoader()
    
    # Test retrieval
    guidelines = loader.get_relevant_guidelines(
        ui_context="mobile checkout page with payment form",
        persona_name="senior_casual",
        top_k=5
    )
    
    print("\n=== Relevant Guidelines ===")
    for i, result in enumerate(guidelines, 1):
        h = result['heuristic']
        print(f"\n{i}. [{h['id']}] {h['title']}")
        print(f"   Source: {h['source']} | Category: {h['category']}")
        print(f"   Similarity: {result['similarity_score']:.2f} | Weighted: {result['weighted_score']:.2f}")
        print(f"   {h['description'][:100]}...")

File: lib/memory/heuristicLoader.ts
/**
 * HeuristicLoader - Persona-weighted UX guideline retrieval
 * 
 * This class provides evidence-based UX guidelines (Baymard, NN/g, WCAG)
 * weighted by persona characteristics. The guidelines SUPPORT persona-based
 * testing, they don't replace it.
 */

export interface UXHeuristic {
  id: string
  source: 'baymard' | 'nng' | 'wcag'
  category: string
  title: string
  description: string
  context: string[] // ['checkout', 'forms', 'mobile']
  severityWeight: number // 0.0-1.0
  personaRelevance: Record<string, number> // persona_name -> relevance score
  examples: string[]
  citationUrl: string
}

export interface PersonaProfile {
  name: string
  age: number
  tech_literacy: string
  eyesight: string
  cognitive_load: string
  attention_rules: string[]
}

export interface WeightedGuideline extends UXHeuristic {
  relevanceScore: number // How relevant to current persona
  personaImpactReason: string // WHY this matters for this persona
}

export class HeuristicLoader {
  private heuristics: UXHeuristic[]

  constructor() {
    this.heuristics = this.initializeDefaultHeuristics()
  }

  /**
   * Get guidelines relevant to the current context and persona.
   * Returns guidelines sorted by persona relevance, not just generic importance.
   */
  async getRelevantGuidelines(
    pageContext: string,
    persona: PersonaProfile,
    topK: number = 5
  ): Promise<WeightedGuideline[]> {
    // Extract context keywords from page
    const contextKeywords = this.extractContextKeywords(pageContext)
    
    // Score each heuristic by persona relevance
    const scored = this.heuristics.map(heuristic => {
      const personaScore = this.calculatePersonaRelevance(heuristic, persona)
      const contextScore = this.calculateContextRelevance(heuristic, contextKeywords)
      const combinedScore = (personaScore * 0.7) + (contextScore * 0.3) // Persona-first weighting
      
      return {
        ...heuristic,
        relevanceScore: combinedScore,
        personaImpactReason: this.generatePersonaImpactReason(heuristic, persona)
      }
    })

    // Sort by relevance and return top K
    return scored
      .sort((a, b) => b.relevanceScore - a.relevanceScore)
      .slice(0, topK)
  }

  /**
   * Calculate how relevant a guideline is to a specific persona.
   * This is where the "human testing" magic happens.
   */
  private calculatePersonaRelevance(
    heuristic: UXHeuristic,
    persona: PersonaProfile
  ): number {
    // Check if persona has explicit relevance score
    const baseScore = heuristic.personaRelevance[persona.name] || 0.5

    // Boost score based on persona characteristics
    let adjustedScore = baseScore

    // Age-based adjustments
    if (persona.age >= 65) {
      if (heuristic.category === 'Touch Targets') adjustedScore += 0.2
      if (heuristic.category === 'Text Readability') adjustedScore += 0.3
      if (heuristic.category === 'Cognitive Load') adjustedScore += 0.2
    }

    // Eyesight-based adjustments
    if (persona.eyesight.includes('low') || persona.eyesight.includes('presbyopia')) {
      if (heuristic.category === 'Contrast') adjustedScore += 0.3
      if (heuristic.category === 'Text Readability') adjustedScore += 0.3
      if (heuristic.category === 'Visual Hierarchy') adjustedScore += 0.2
    }

    // Tech literacy adjustments
    if (persona.tech_literacy === 'low') {
      if (heuristic.category === 'Navigation') adjustedScore += 0.2
      if (heuristic.category === 'Error Prevention') adjustedScore += 0.3
      if (heuristic.category === 'Microcopy') adjustedScore += 0.2
    }

    // Cognitive load adjustments
    if (persona.cognitive_load === 'low') {
      if (heuristic.category === 'Cognitive Load') adjustedScore += 0.3
      if (heuristic.category === 'Information Density') adjustedScore += 0.2
    }

    return Math.min(adjustedScore, 1.0)
  }

  /**
   * Calculate how relevant a guideline is to the current page context
   */
  private calculateContextRelevance(
    heuristic: UXHeuristic,
    contextKeywords: string[]
  ): number {
    const matches = heuristic.context.filter(ctx => 
      contextKeywords.some(keyword => 
        ctx.toLowerCase().includes(keyword.toLowerCase()) ||
        keyword.toLowerCase().includes(ctx.toLowerCase())
      )
    )
    
    return matches.length / Math.max(heuristic.context.length, 1)
  }

  /**
   * Extract relevant keywords from page context
   */
  private extractContextKeywords(pageContext: string): string[] {
    const keywords: string[] = []
    const lower = pageContext.toLowerCase()

    // Page type detection
    if (lower.includes('checkout') || lower.includes('cart')) keywords.push('checkout')
    if (lower.includes('form') || lower.includes('input')) keywords.push('forms')
    if (lower.includes('product') || lower.includes('price')) keywords.push('product')
    if (lower.includes('navigation') || lower.includes('menu')) keywords.push('navigation')
    if (lower.includes('search')) keywords.push('search')
    if (lower.includes('button')) keywords.push('buttons')
    if (lower.includes('link')) keywords.push('links')
    if (lower.includes('mobile') || lower.includes('responsive')) keywords.push('mobile')

    return keywords
  }

  /**
   * Generate a human-readable explanation of why this guideline matters for this persona
   */
  private generatePersonaImpactReason(
    heuristic: UXHeuristic,
    persona: PersonaProfile
  ): string {
    const reasons: string[] = []

    // Age-based reasoning
    if (persona.age >= 65) {
      if (heuristic.category === 'Touch Targets') {
        reasons.push(`Seniors often have reduced motor control and may struggle with small tap targets`)
      }
      if (heuristic.category === 'Text Readability') {
        reasons.push(`Age-related vision changes make small or low-contrast text difficult to read`)
      }
    }

    // Eyesight-based reasoning
    if (persona.eyesight.includes('low') || persona.eyesight.includes('presbyopia')) {
      if (heuristic.category === 'Contrast') {
        reasons.push(`Users with ${persona.eyesight} require higher contrast ratios to distinguish elements`)
      }
      if (heuristic.category === 'Text Readability') {
        reasons.push(`Visual impairments make font size and spacing critical for readability`)
      }
    }

    // Tech literacy reasoning
    if (persona.tech_literacy === 'low') {
      if (heuristic.category === 'Navigation') {
        reasons.push(`Users with low tech literacy need clear, obvious navigation patterns`)
      }
      if (heuristic.category === 'Error Prevention') {
        reasons.push(`Less tech-savvy users benefit from proactive error prevention`)
      }
    }

    // Cognitive load reasoning
    if (persona.cognitive_load === 'low') {
      if (heuristic.category === 'Cognitive Load') {
        reasons.push(`Users with limited cognitive capacity need simplified, focused interfaces`)
      }
    }

    return reasons.length > 0 
      ? reasons.join('. ') 
      : `This guideline applies to ${persona.name} based on general UX best practices`
  }

  /**
   * Initialize curated heuristics database
   * Focus: Persona-relevant guidelines, not exhaustive UX catalog
   */
  private initializeDefaultHeuristics(): UXHeuristic[] {
    return [
      // === BAYMARD INSTITUTE - E-COMMERCE FOCUSED ===
      {
        id: 'BAY-001',
        source: 'baymard',
        category: 'Touch Targets',
        title: 'Minimum Touch Target Size',
        description: 'Touch targets should be at least 44x44 CSS pixels to ensure easy tapping, especially for users with motor impairments or on mobile devices.',
        context: ['mobile', 'buttons', 'links', 'forms'],
        severityWeight: 0.9,
        personaRelevance: {
          'senior_casual': 1.0,
          'accessibility_focused': 1.0,
          'middle_age_moderate': 0.7,
          'young_power_user': 0.3
        },
        examples: [
          'Button too small: 28x28px',
          'Link spacing insufficient: 8px between targets'
        ],
        citationUrl: 'https://baymard.com/blog/touch-target-size'
      },
      {
        id: 'BAY-002',
        source: 'baymard',
        category: 'Form Design',
        title: 'Inline Form Validation',
        description: 'Provide immediate, inline validation feedback as users complete form fields. Show success indicators for valid input and clear error messages for invalid input.',
        context: ['forms', 'checkout', 'registration'],
        severityWeight: 0.8,
        personaRelevance: {
          'senior_casual': 0.9,
          'accessibility_focused': 0.8,
          'middle_age_moderate': 0.8,
          'young_power_user': 0.6
        },
        examples: [
          'No validation until form submission',
          'Error messages appear only at top of page'
        ],
        citationUrl: 'https://baymard.com/blog/inline-form-validation'
      },
      {
        id: 'BAY-003',
        source: 'baymard',
        category: 'Checkout',
        title: 'Guest Checkout Option',
        description: 'Always offer a guest checkout option. Forced account creation is a major cause of cart abandonment, especially for first-time or infrequent shoppers.',
        context: ['checkout', 'registration'],
        severityWeight: 0.9,
        personaRelevance: {
          'senior_casual': 0.9,
          'accessibility_focused': 0.6,
          'middle_age_moderate': 0.8,
          'young_power_user': 0.7
        },
        examples: [
          'Forced account creation before purchase',
          'No clear "Continue as Guest" button'
        ],
        citationUrl: 'https://baymard.com/blog/guest-checkout'
      },
      {
        id: 'BAY-004',
        source: 'baymard',
        category: 'Navigation',
        title: 'Clear Category Labels',
        description: 'Use plain language for navigation labels. Avoid marketing jargon, clever wordplay, or ambiguous terms that require users to guess what they mean.',
        context: ['navigation', 'menu', 'product'],
        severityWeight: 0.7,
        personaRelevance: {
          'senior_casual': 1.0,
          'accessibility_focused': 0.8,
          'middle_age_moderate': 0.7,
          'young_power_user': 0.4
        },
        examples: [
          'Menu item labeled "Lifestyle" instead of "Clothing"',
          'Ambiguous category names like "Essentials"'
        ],
        citationUrl: 'https://baymard.com/blog/category-names'
      },

      // === NIELSEN NORMAN GROUP - USABILITY FOCUSED ===
      {
        id: 'NNG-001',
        source: 'nng',
        category: 'Text Readability',
        title: 'Readable Font Size',
        description: 'Body text should be at least 16px on desktop and mobile. Smaller text creates eye strain and is particularly difficult for older users or those with vision impairments.',
        context: ['text', 'content', 'mobile'],
        severityWeight: 0.8,
        personaRelevance: {
          'senior_casual': 1.0,
          'accessibility_focused': 1.0,
          'middle_age_moderate': 0.8,
          'young_power_user': 0.4
        },
        examples: [
          'Body text at 12px or 14px',
          'Fine print at 10px or smaller'
        ],
        citationUrl: 'https://www.nngroup.com/articles/font-size/'
      },
      {
        id: 'NNG-002',
        source: 'nng',
        category: 'Error Prevention',
        title: 'Prevent Errors Before They Occur',
        description: 'Design interfaces to prevent errors rather than just detecting them. Use constraints, defaults, and clear affordances to guide users toward correct actions.',
        context: ['forms', 'interaction', 'buttons'],
        severityWeight: 0.9,
        personaRelevance: {
          'senior_casual': 1.0,
          'accessibility_focused': 0.9,
          'middle_age_moderate': 0.8,
          'young_power_user': 0.5
        },
        examples: [
          'No input masking for phone/credit card fields',
          'Destructive actions without confirmation dialogs'
        ],
        citationUrl: 'https://www.nngroup.com/articles/slips/'
      },
      {
        id: 'NNG-003',
        source: 'nng',
        category: 'Contrast',
        title: 'Sufficient Color Contrast',
        description: 'Text and interactive elements must have sufficient contrast against their background. Minimum 4.5:1 for normal text, 3:1 for large text (WCAG AA).',
        context: ['text', 'buttons', 'links', 'accessibility'],
        severityWeight: 0.9,
        personaRelevance: {
          'senior_casual': 1.0,
          'accessibility_focused': 1.0,
          'middle_age_moderate': 0.8,
          'young_power_user': 0.5
        },
        examples: [
          'Light gray text on white background',
          'Low-contrast buttons that blend into page'
        ],
        citationUrl: 'https://www.nngroup.com/articles/low-contrast/'
      },
      {
        id: 'NNG-004',
        source: 'nng',
        category: 'Cognitive Load',
        title: 'Minimize Cognitive Load',
        description: 'Reduce the amount of information users must process at once. Break complex tasks into steps, use progressive disclosure, and avoid overwhelming users with choices.',
        context: ['forms', 'checkout', 'navigation'],
        severityWeight: 0.8,
        personaRelevance: {
          'senior_casual': 1.0,
          'accessibility_focused': 0.8,
          'middle_age_moderate': 0.9,
          'young_power_user': 0.4
        },
        examples: [
          'Single-page form with 30+ fields',
          'Too many navigation options (>7 items)'
        ],
        citationUrl: 'https://www.nngroup.com/articles/minimize-cognitive-load/'
      },

      // === WCAG 2.1 - ACCESSIBILITY FOCUSED ===
      {
        id: 'WCAG-001',
        source: 'wcag',
        category: 'Keyboard Navigation',
        title: 'Keyboard Accessible',
        description: 'All interactive elements must be accessible via keyboard alone. Users should be able to navigate, activate, and interact without a mouse.',
        context: ['accessibility', 'forms', 'buttons', 'navigation'],
        severityWeight: 1.0,
        personaRelevance: {
          'senior_casual': 0.6,
          'accessibility_focused': 1.0,
          'middle_age_moderate': 0.5,
          'young_power_user': 0.7
        },
        examples: [
          'Custom dropdowns not keyboard accessible',
          'Modal dialogs that trap focus'
        ],
        citationUrl: 'https://www.w3.org/WAI/WCAG21/Understanding/keyboard.html'
      },
      {
        id: 'WCAG-002',
        source: 'wcag',
        category: 'ARIA Labels',
        title: 'Meaningful ARIA Labels',
        description: 'Interactive elements must have descriptive ARIA labels or accessible names. Screen reader users rely on these to understand element purpose.',
        context: ['accessibility', 'buttons', 'links', 'forms'],
        severityWeight: 1.0,
        personaRelevance: {
          'senior_casual': 0.4,
          'accessibility_focused': 1.0,
          'middle_age_moderate': 0.3,
          'young_power_user': 0.3
        },
        examples: [
          'Icon buttons with no aria-label',
          'Generic "Click here" link text'
        ],
        citationUrl: 'https://www.w3.org/WAI/WCAG21/Understanding/name-role-value.html'
      },
      {
        id: 'WCAG-003',
        source: 'wcag',
        category: 'Visual Hierarchy',
        title: 'Proper Heading Structure',
        description: 'Use semantic HTML headings (h1-h6) in logical order. This helps all users understand page structure, especially screen reader users.',
        context: ['accessibility', 'content', 'navigation'],
        severityWeight: 0.7,
        personaRelevance: {
          'senior_casual': 0.6,
          'accessibility_focused': 1.0,
          'middle_age_moderate': 0.5,
          'young_power_user': 0.3
        },
        examples: [
          'Skipping heading levels (h1 to h3)',
          'Using <div> with CSS instead of semantic headings'
        ],
        citationUrl: 'https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships.html'
      }
    ]
  }
}
